File: leetcode/763_partition_labels.cpp
#include <cassert>
#include <map>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> partitionLabels (string S) {
        map<char, int> lastOccurrences;
        for (size_t i = 0; i < S.length(); i++) {
            lastOccurrences[S[i]] = i;
        }

        vector<int> result;
        size_t currentCharLast, currentFirst = 0, currentLast = 0;

        for (size_t i = 0; i < S.length(); i++) {
            currentCharLast = lastOccurrences[S[i]];
            currentLast = max(currentLast, currentCharLast);
            if (i == currentLast) {
                result.push_back(i - currentFirst + 1);
                currentFirst = i + 1;
            }
        }

        return result;
    }
};

int main () {
    auto solution = Solution();

    vector<int> expected = {9, 7, 8};
    vector<int> result = solution.partitionLabels("ababcbacadefegdehijhklij");

    assert(expected == result);
}

File: leetcode/763_partition_labels.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function


class Solution(object):
    def partitionLabels(self, S):
        last_occurrences = {c: i for i, c in enumerate(S)}

        result = []
        current_first, current_last = 0, 0

        for i, c in enumerate(S):
            current_last = max(current_last, last_occurrences[c])
            if i == current_last:
                result.append(i - current_first + 1)
                current_first = i + 1

        return result


if __name__ == '__main__':
    solution = Solution()

    assert [9, 7, 8] == solution.partitionLabels('ababcbacadefegdehijhklij')
